clang/lib/CIR/CodeGen/CIRGenExprConstant.cpp

```diff
@@ -365,10 +365,15 @@ mlir::Attribute ConstantEmitter::tryEmitPrivateForVarInit(const VarDecl &d) {
   if (!d.hasLocalStorage()) {
     QualType ty = cgm.getASTContext().getBaseElementType(d.getType());
     if (ty->isRecordType())
-      if (d.getInit() && isa<CXXConstructExpr>(d.getInit())) {
-        cgm.errorNYI(d.getInit()->getBeginLoc(),
-                     "tryEmitPrivateForVarInit CXXConstructExpr");
-        return {};
+      if (const CXXConstructExpr *e =
+              dyn_cast_or_null<CXXConstructExpr>(d.getInit())) {
+        const CXXConstructorDecl *cd = e->getConstructor();
+        // FIXME: we should probably model this more closely to C++ than
+        // just emitting a global with zero init (mimic what we do for trivial
+        // assignments and whatnots). Since this is for globals shouldn't
+        // be a problem for the near future.
+        if (cd->isTrivial() && cd->isDefaultConstructor())
+          return cir::ZeroAttr::get(cgm.convertType(d.getType()));
```

Comment:
OGCG calls `emitNullConstant` here. We have `CIRGenModule::emitNullConstant` partially implemented. In the incubator `CIRGenModule::emitNullConstant` fails with a NYi if the record type is not zero initializable and IMO that's a reasonable limitation to have for now (instead of silently breaking vtables and default ctors).

Reply:
Yeah, that seems reasonable.

Reply:
As I tried to implement this suggestion, it quickly came to my attention that `emitNullConstant` returns a value, and we need an attribute here. I think the `cd->isTrivial() && cd->isDefaultConstructor()` check should prevent us from getting here in any case where the zero attribute isn't sufficient, but I will add an assert that the type is zero-initializable.

Reply:
Second complication, the clang AST says that `CompleteS` is not zero-initializable for the test case I'm adding in struct.cpp. (But classic codegen generates `zeroinitializer` for it). I think I may need to implement `CIRGenRecordLayout::isZeroInitializableAsBase` for this.

Reply:
Thanks for adding the `errorNYI` for structs that are not "truly" zero initializable for a given ABI. LGTM now.